wip: create descriptors for public schema

RichardJCai opened the pull request.

RichardJCai committed:
Bootstrap new databases with public schemas that have descriptors.
Version gating.

Comment about version gating. Separate out backup/restore variable for future updates. Add todos.

Release justification: None, waiting for 22.1
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68674)
<!-- Reviewable:end -->

postamar commented on pkg/sql/catalog/bootstrap/metadata.go:
```diff
@@ -115,21 +115,19 @@ func (ms MetadataSchema) GetInitialValues() ([]roachpb.KeyValue, []roachpb.RKey)
-					Key:   catalogkeys.MakePublicSchemaNameKey(ms.codec, desc.GetID()),
-					Value: publicSchemaValue,
-				})
+			)
```
This whole if/else block can be replaced with
```
ret = append(ret, roachpb.KeyValue{Key: catalogkeys.EncodeNameKey(ms.codec, desc), Value: value})
```
as we now have a 1-to-1 mapping from descriptor to namespace entry.

postamar reviewed (commented):
Not sure if this is helpful but here's what I thought a migration would look like:
https://gist.github.com/postamar/bd11f1abe032eb6887d78643c94ad927
https://gist.github.com/postamar/b9da9ef2f3714b4dd0f1510900074365

I don't have anything else that's meaningful to share, I'm afraid. The migration does the change using the catalog API but I've been wondering if it would be better to do it via SQL statements, as in: for each database create a new `crdb_internal_public_migration` schema or something and reparent everything from `public` to that new schema using `ALTER TABLE ... SET SCHEMA` and then at the very end rename that schema to `public` somehow.
I'm sorry this is vague but I didn't go further because I was stumped by the bootstrapping as well. I got a `WriteTooOldError` but I hadn't added zone configs for the new system schema descriptor so perhaps I did it wrong. For what it's worth I also thought that using ID `29` for the system schema was a good idea.

RichardJCai commented:
I think I've managed to figure out the bootstrapping issue.

Now another question I have is if a startup migration is better in this case, version gating all the schema resolution logic seems tedious.

ajwerner commented:
> Now another question I have is if a startup migration is better in this case, version gating all the schema resolution logic seems tedious.

In principle I'm opposed to the startup migration. We have so many things which scan the namespace table and then look up descriptors and that code is not expecting a desc. Counting on them doing the right thing seems hard. 

Also it's an O(descriptors) migration. So, yeah, generally I don't love that idea. The flip side is, as you pointed out, figuring out all the spots we treat it differently or assume things and combining version gating with other techniques.

RichardJCai committed:
sql: disallow alter table rename moving a table to another database

Moving a table from one database's public schema to another was
previously possible using ALTER TABLE RENAME due to the public
schema being a special case, now that the public schema is a
regular user defined schema with a descriptor, this is no
longer the case.

Release note (sql change): Previously, ALTER TABLE ... RENAME TO ...
would allow the user to move the table from a database to another
if the table is being moved within one database's public schema
to another. This is now disallowed.

RichardJCai committed:
sql: fix tests using 53 as first table id

Release note: None

RichardJCai committed:
Add HasPublicSchemaWithDescriptor function to database descriptors

RichardJCai committed:
sql: explicitly disallow dropping public schema and reassigning public schema owner

This maintains behaviour from 21.2 and prior. We can revisit this in 22.1 and remove these restrictions

Release justification: None
Release note: None

RichardJCai committed:
rttanalysis: update benchmark expectations to consider public schema has a descriptor

Release note: None

Release justification:

RichardJCai committed:
sql: disallow alter database convert to schema on versions with explicit public schemas

Release note (sql change): ALTER DATABASE CONVERT TO SCHEMA is now disabled on
versions 22.1+.

RichardJCai committed:
cli: update TestDoctorZipDir and TestDoctorCluster test

Release justification: None, waiting for 22.1.
Release note: None

RichardJCai committed:
ccl, bulkio: skip IMPORT TABLE tests

Import fails due to the public schema becoming a regular
schema with a descriptor.

These tests are skipped currently for clarity and
hopefully in preparation for import to become fully removed.

Release justification: None, waiting for 22.1
Release note: None

RichardJCai committed:
backup/restore work

RichardJCai committed:
restore: fix full cluster restore

Fix full cluster restore now that public schemas are backed
by descriptors.

Release note: None

RichardJCai committed:
backup restore test fixes

Release justification:
Release note (<category, see below>): <what> <show> <why>

RichardJCai committed:
backup/restore: fix multiregion backup/restore

Release justification: None, waiting for 22.1
Release note: None

RichardJCai committed:
Fix restoring table that uses a type that is already defined in the schema being restored to

RichardJCai committed:
More backup/restore fixes
